core/src/main/java/org/stellar/anchor/sep10/Sep10Service.java

```diff
@@ -58,167 +57,248 @@ public Sep10Service(
-                  new Network(appConfig.getStellarNetworkPassphrase()),
-                  challengeRequest.getAccount(),
-                  challengeRequest.getHomeDomain(),
-                  sep10Config.getWebAuthDomain(),
-                  new TimeBounds(now, now + sep10Config.getAuthTimeout()),
-                  (challengeRequest.getClientDomain() == null)
-                      ? ""
-                      : challengeRequest.getClientDomain(),
-                  (clientSigningKey == null) ? "" : clientSigningKey,
-                  memo);
-      // Convert the challenge to response
-      trace("SEP-10 challenge txn:", txn);
-      ChallengeResponse challengeResponse =
-          ChallengeResponse.of(txn.toEnvelopeXdrBase64(), appConfig.getStellarNetworkPassphrase());
-      trace("challengeResponse:", challengeResponse);
-      // increment counter
-      sep10ChallengeCreatedCounter.increment();
-      return challengeResponse;
-    } catch (InvalidSep10ChallengeException ex) {
-      warnEx(ex);
-      throw new SepException("Failed to create the sep-10 challenge.", ex);
-    }
+  @Override
+  public void incrementChallengeRequestCreatedCounter() {
+    sep10ChallengeCreatedCounter.increment();
   }
 
-  public ValidationResponse validateChallenge(ValidationRequest validationRequest)
-      throws IOException, InvalidSep10ChallengeException, SepValidationException {
-    if (validationRequest == null || validationRequest.getTransaction() == null) {
-      throw new SepValidationException("{transaction} is required.");
-    }
+  @Override
+  public void preValidateRequestValidation(
+      ValidationRequest request, ChallengeTransaction challenge) {
+    // NOOP. To be overridden.
+  }
 
-    String clientDomain = validateChallenge(validationRequest.getTransaction());
-    return ValidationResponse.of(
-        generateSep10Jwt(validationRequest.getTransaction(), clientDomain));
+  @Override
+  public void incrementValidationRequestValidatedCounter() {
+    sep10ChallengeValidatedCounter.increment();
   }
 
-  public String validateChallenge(String challengeXdr)
-      throws IOException, InvalidSep10ChallengeException {
-    debug("Parse challenge string.");
-    Sep10Challenge.ChallengeTransaction challenge =
-        Sep10ChallengeWrapper.instance()
-            .readChallengeTransaction(
-                challengeXdr,
-                serverAccountId,
-                new Network(appConfig.getStellarNetworkPassphrase()),
-                sep10Config.getHomeDomain(),
-                sep10Config.getWebAuthDomain());
+  String fetchSigningKeyFromClientDomain(String clientDomain) throws SepException {
+    return Sep10Helper.fetchSigningKeyFromClientDomain(clientDomain);
+  }
 
-    debugF(
-        "Challenge parsed. account={}, home_domain={}",
-        shorter(challenge.getClientAccountId()),
-        challenge.getMatchedHomeDomain());
+  void validateHomeDomain(ChallengeRequest request) throws SepValidationException {
```

Comment:
What do you think about moving some of the validations into their own class? We do something like this in [SEP-6](https://github.com/stellar/java-stellar-anchor-sdk/blob/sep-6/core/src/main/java/org/stellar/anchor/sep6/RequestValidator.java). I found that it made the `Sep6Service` easier to test.

Reply:
Defiantly. Do you mind to extract this method in a SEP-6 PR? This way, the use case will be more solid.